include/kernel.h

```diff
@@ -1620,24 +1582,55 @@ __syscall u32_t k_timer_status_get(struct k_timer *timer);
  */
 __syscall u32_t k_timer_status_sync(struct k_timer *timer);
 
-extern s32_t z_timeout_remaining(struct _timeout *timeout);
+extern k_ticks_t z_timeout_end(struct _timeout *timeout);
+extern k_ticks_t z_timeout_remaining(struct _timeout *timeout);
+
+/**
+ * @brief Get uptime expiration of a timer
+ *
+ * This routine returns the value of the system uptime counter at
+ * which point a running timer will expire. If the timer is not
+ * running, it returns K_FOREVER_TICKS.
+ *
+ * @param timer     Address of timer.
+ *
+ * @return Uptime (in ticks).
+ */
+__syscall k_ticks_t k_timer_end_ticks(struct k_timer *timer);
+
+static inline k_ticks_t z_impl_k_timer_end_ticks(struct k_timer *timer)
+{
+	return z_timeout_end(&timer->timeout);
+}
 
 /**
  * @brief Get time remaining before a timer next expires.
  *
- * This routine computes the (approximate) time remaining before a running
- * timer next expires. If the timer is not running, it returns zero.
+ * This routine returns the number of system ticks that will elapse
+ * before a running timer next expires (that is, the time until
+ * expiration is at least as long as the return value). If the timer
+ * is not running, it returns K_FOREVER_TICKS.
  *
  * @param timer     Address of timer.
  *
- * @return Remaining time (in milliseconds).
+ * @return Remaining time (in ticks).
  */
-__syscall u32_t k_timer_remaining_get(struct k_timer *timer);
+__syscall k_ticks_t k_timer_remaining_ticks(struct k_timer *timer);
 
-static inline u32_t z_impl_k_timer_remaining_get(struct k_timer *timer)
+static inline k_ticks_t z_impl_k_timer_remaining_ticks(struct k_timer *timer)
 {
-	const s32_t ticks = z_timeout_remaining(&timer->timeout);
-	return (ticks > 0) ? (u32_t)__ticks_to_ms(ticks) : 0U;
+	return z_timeout_remaining(&timer->timeout);
```

Comment:
Experimentation with the test application in https://github.com/pabigot/zephyr/commits/analyze/17155 shows that this sometimes returns negative numbers, which results in invalid return values from `k_timer_remaining_get()`:
```
Starting...
        T96 S1 rem 131070876 ms -36815 tick
... done
```